### PR TITLE
Add Vec::get_uninit_unchecked

### PR DESCRIPTION
**Pull request by moshg**
This is intended to reduce misusing `get_unchecked_mut` to access out of bounds.
In addition, we can detect out-of-allocation with this method unlike `x.as_ptr_mut().add(index)`.

I hope this to help `get_unchecked` to do bound check in debug build in the future. https://github.com/rust-lang/rust/issues/36976

Though immutable `get_uninit_unchecked` is useless, should I name `get_uninit_unchecked_mut`?

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -854,6 +854,42 @@ impl<T> Vec<T> {
...
+    ///     for i in 0..size {
+    ///         x.get_uninit_unchecked(i).write(i as i32);
+    ///     }
+    ///     x.set_len(size);
+    /// }
+    /// assert_eq!(&*x, &[0, 1, 2, 3]);
+    /// ```
+    #[unstable(feature = "vec_get_uninit_unchecked", issue = "none")]
+    pub fn get_uninit_unchecked(&mut self, index: usize) -> &mut MaybeUninit<T> {
+        if cfg!(debug_assertions) && index >= self.capacity() {
```
Comment (Centril) [inline src/liballoc/vec.rs:886]: Use a `debug_assert!` instead.

File: src/liballoc/vec.rs [hunk 2]
```diff
@@ -854,6 +854,42 @@ impl<T> Vec<T> {
...
+    /// unsafe {
+    ///     for i in 0..size {
+    ///         x.get_uninit_unchecked(i).write(i as i32);
+    ///     }
+    ///     x.set_len(size);
+    /// }
+    /// assert_eq!(&*x, &[0, 1, 2, 3]);
+    /// ```
+    #[unstable(feature = "vec_get_uninit_unchecked", issue = "none")]
+    pub fn get_uninit_unchecked(&mut self, index: usize) -> &mut MaybeUninit<T> {
```
Comment (Centril) [inline src/liballoc/vec.rs:885]:
```suggestion
    pub unsafe fn get_uninit_unchecked(&mut self, index: usize) -> &mut MaybeUninit<T> {
```

There are two ways in which you can violate soundness with this method:

1. You provide `index` where `index >= self.capacity()`.
2. You do `let elem = vec.get_uninit_unchecked(index); *elem = MaybeUninit::uninit(); drop(vec[index]);`.

Comment (Centril) [reply src/liballoc/vec.rs:885]: Also, you can have a memory leak if you forget to use `vec.set_len(...)` afterwards in some circumstances; also worth noting in docs.

**Comment by joshtriplett** [comment]
Tagging libs, and assigning to a random libs team member.

**Comment by BurntSushi** [comment]
I'm generally not on the PR review rotation. r? @sfackler 

Also, cc @RalfJung

**Comment by moshg** [comment]
@Centril Thank you for your review!

I cannot find broken links. How can I find them?

With `python x.py doc` no error occurs. What command do I have to do to get the same error as CI?

**Comment by sfackler** [comment]
Rather than conflating unchecked indexing and conversion to `MaybeUninit`, how about a method like this?

```rust
impl<T> Vec<T> {
    /// Returns the vector's spare capacity as a mutable slice of maybe-uninitialized values.
    ///
    /// The length of the slice will be `self.capacity() - self.len()`.
    pub fn spare_capacity_mut(&mut self) -> &mut [MaybeUninit<T>] { ... }
}
```

**Comment by joelpalmer** [comment]
Ping from Triage: Hi @moshg - any updates?

**Comment by joelpalmer** [comment]
Ping from Triage: Closing due to inactivity. @moshg please reopen with updates. Thanks for the PR.

**Pull request closed without merging**